hoomd/hpmc/ExternalFieldHarmonic.h

```diff
@@ -0,0 +1,468 @@
+        {
+        return m_k_rotational;
+        }
+
+    //! Helper function to be called when particles are added/removed
+    void slotGlobalParticleNumChange()
+        {
+        if (m_reference_positions.size() != this->m_pdata->getNGlobal()
+            || m_reference_orientations.size() != this->m_pdata->getNGlobal())
+            {
+            throw std::runtime_error("Number of particles no longer equals number of reference "
+                                     "points in ExternalFieldHarmonic.");
+            }
+        }
+
+    /** Calculate the change in energy for trial moves
+     *
+     * This function currently ignores any information associated with box changes.
+     * However, this function only gets called in the box updater, so it should not ignore box
+     * changes. But why would you have a lattice field and a box updater active in the same
+     * simulation?
+     */
+    double calculateDeltaE(const Scalar4* const position_old_arg, // why is this a Scalar4?
+                           const Scalar4* const orientation_old_arg,
+                           const BoxDim* const box_old_arg)
+        {
+        ArrayHandle<Scalar4> h_pos(m_pdata->getPositions(),
+                                   access_location::host,
+                                   access_mode::readwrite);
+        ArrayHandle<Scalar4> h_orient(m_pdata->getOrientationArray(),
+                                      access_location::host,
+                                      access_mode::readwrite);
+        const Scalar4* const position_new = h_pos.data; // current positions from system definition
+        const Scalar4* const orientation_new
+            = h_orient.data; // current orientations from system definition
+        const Scalar4 *position_old = position_old_arg, *orientation_old = orientation_old_arg;
+        if (!position_old)
+            position_old = position_new;
+        if (!orientation_old)
+            orientation_old = orientation_new;
+
+        double dE = 0.0;
+        for (unsigned int i = 0; i < m_pdata->getN(); i++)
+            {
+            Scalar old_E
+                = calcE(i, vec3<Scalar>(*(position_old + i)), quat<Scalar>(*(orientation_old + i)));
+            Scalar new_E
+                = calcE(i, vec3<Scalar>(*(position_new + i)), quat<Scalar>(*(orientation_new + i)));
+            dE += new_E - old_E;
+            }
+
+#ifdef ENABLE_MPI
+        if (this->m_sysdef->isDomainDecomposed())
+            {
+            MPI_Allreduce(MPI_IN_PLACE,
+                          &dE,
+                          1,
+                          MPI_HOOMD_SCALAR,
+                          MPI_SUM,
+                          m_exec_conf->getMPICommunicator());
+            }
+#endif
+
+        return dE;
+        }
+
+    /** This function gets called during the update step in the MC integrator
+     *
+     * What purpose does this function serve?
+     */
+    Scalar getEnergy(uint64_t timestep)
```

Comment:
Would users be interested in accessing energy components separately in `energy_translational` and `energy_rotational`? If so, we could add these properties. Keep the `energy` property that computes the total.

Reply:
Yes, I think users would be interested in something like that. I haven't worked out the details but I imagine this would be useful for getting rotational and translational contributions to the free energy somehow. And it's easy enough to add, so I'll put it in.